src/test/client/nonblocking.cc

```diff
@@ -208,3 +208,87 @@ TEST_F(TestClient, LlreadvLlwritevNullContext) {
   client->ll_release(fh);
   ASSERT_EQ(0, client->ll_unlink(root, filename, myperm));
 }
+
+TEST_F(TestClient, LlreadvLlwritevInvalidFileHandle) {
+  /* Test provding null or invalid file handle returns an error
+  as expected*/
+
+  Fh *fh_null = NULL;
+
+  char out_buf_0[] = "hello ";
+  char out_buf_1[] = "world\n";
+  struct iovec iov_out[2] = {
+    {out_buf_0, sizeof(out_buf_0)},
+    {out_buf_1, sizeof(out_buf_1)},
+  };
+
+  char in_buf_0[sizeof(out_buf_0)];
+  char in_buf_1[sizeof(out_buf_1)];
+  struct iovec iov_in[2] = {
+    {in_buf_0, sizeof(in_buf_0)},
+    {in_buf_1, sizeof(in_buf_1)},
+  };
+
+  std::unique_ptr<C_SaferCond> writefinish = nullptr;
+  std::unique_ptr<C_SaferCond> readfinish = nullptr;
+
+  writefinish.reset(new C_SaferCond("test-nonblocking-writefinish-invalid-fd"));
+  readfinish.reset(new C_SaferCond("test-nonblocking-readfinish-invalid-fd"));
+
+  int64_t rc;
+  bufferlist bl;
+  ssize_t bytes_written = 0, bytes_read = 0;
+
+  rc = client->ll_preadv_pwritev(fh_null, iov_out, 2, 0, true,
+                                 writefinish.get(), nullptr);
+  ASSERT_EQ(rc, 0);
+  bytes_written = writefinish->wait();
+  ASSERT_EQ(bytes_written, -CEPHFS_EBADF);
+
+  rc = client->ll_preadv_pwritev(fh_null, iov_in, 2, 0, true,
+                                 readfinish.get(), &bl);
+  ASSERT_EQ(rc, 0);
+  bytes_read = readfinish->wait();
+  ASSERT_EQ(bytes_read, -CEPHFS_EBADF);
+  ASSERT_EQ(bl.length(), 0);
+
+  // test after closing the file handle
+  int mypid = getpid();
+  char filename[256];
+
+  client->unmount();
+  TearDown();
+  SetUp();
+
+  sprintf(filename, "test_llreadvllwritevclosedfdfile%u", mypid);
+
+  Inode *root, *file;
+  root = client->get_root();
+  ASSERT_NE(root, (Inode *)NULL);
+
+  Fh *fh;
+  struct ceph_statx stx;
+
+  ASSERT_EQ(0, client->ll_createx(root, filename, 0666,
+				  O_RDWR | O_CREAT | O_TRUNC,
+				  &file, &fh, &stx, 0, 0, myperm));
+
+  client->ll_release(fh);
+  ASSERT_EQ(0, client->ll_unlink(root, filename, myperm));
+
+  writefinish.reset(new C_SaferCond("test-nonblocking-writefinish-closed-fd"));
+  readfinish.reset(new C_SaferCond("test-nonblocking-readfinish-closed-fd"));
+
+  rc = client->ll_preadv_pwritev(fh, iov_out, 2, 0, true, writefinish.get(),
+                                 nullptr);
+  ASSERT_EQ(rc, 0);
+  bytes_written = writefinish->wait();
+  ASSERT_EQ(bytes_written, -CEPHFS_EBADF);
+
+  rc = client->ll_preadv_pwritev(fh, iov_in, 2, 0, true, readfinish.get(),
+                                 &bl);
```

Comment:
Presuming that this is a READ op, shouldn't the 5th parameter be `false` instead ?

Reply:
right, fixed it. thanks